zyte_common_items/pages.py

```diff
@@ -0,0 +1,53 @@
+from datetime import datetime
+
+import attrs
+from web_poet import ItemPage, ResponseUrl, Returns, WebPage, field
+
+from .components import Metadata
+from .items import Product, ProductList
+
+
+@attrs.define
+class BasePage(ItemPage):
+    _url: ResponseUrl
```

Comment:
Imagine you write a page object class that needs a _request_ URL, e.g. extracts some product ID from the request URL, and uses an additional request to fetch the product data from a JSON API.

For that scenario, subclasssing `BasePage` directly or indirectly will cause your page object class to get an extra, unnecessary `ResponseUrl` input, which requires an actual request (or more) to resolve any redirect from `RequestUrl`.

While I don’t think this should be a blocker here, we might want to try and think how we want to address this eventually.

For example, we could eventually have some method in `ItemPage` that allows querying for an input that may not exist, and then we can remove `_url` from here, so that `ResponseUrl` no longer becomes a required input, and redefine the url field in a way that uses any `ResponseUrl` if defined, or raises `NotImplementedError` otherwise:

```python
    @field
    def url(self) -> str:
        try:
            url = self.get_input(ResponseUrl)
        except ValueError:
            raise NotImplementedError
        return str(url)
```

Note: I assume `self.get_input(ResponseUrl)` would work if e.g. `HttpResponse` exists as input, since `ResponseUrl` can be built from it.

A page object class without a `ResponseUrl`-compatible input would have to reimplement the field to return the right URL, or else get `NotImplementedError` raised when `to_item()` is called.

Reply:
I think that could be a big issue. 

It seems there are 2 sides of this issue. 

First is related to the problem at hand. It seems that using RequestUrl to set Product.url may be not precise enough, as product.url should contain response url. But even if it's not precise, I'm not sure the alternative is better (ask users to provide real url which was used when they make a request using HttpClient). Take an example of AJAX request. You don't want URL of the API endpoint in the "url" field: it is weird for the end users, leaking internal details, and also it means that product can't be extracted by the page object if product.url is passed to it as an input, which is against the field definition. So, I think that RequestUrl is a relatively safe choice. 

We might even consider using it as a default for everything. It may be a better alternative than a trap with ResponseUrl triggering download.

Second is a more general issue, which affects more than RequestUrl vs ResponseUrl. There are other cases when there is something which is expensive to get, and something which is easy to get, and you want an expensive version only if it's "free", i.e. you need to do an expensive operation anyways.

Examples:

* ResponseUrl vs RequestUrl
* ResponseUrl from HttpResponse vs url from BrowserHtml (not sure about that, we don't have a separate url for browser now)
* sometimes code is written to work both with httpResponseBody and with browserHtml, and you want to get something which is already available

In andi there is support for [Unions](https://github.com/scrapinghub/andi#union) and [Optional](https://github.com/scrapinghub/andi#optional), so it's possible to write

```py
any_html: BrowserHtml | HttpResponse
response_url: Optional[ResponseUrl]
```

The semantics is "use BrowserHtml if available; if not, use HttpResponse". The issue is with a definition of "available" - it's defined as "can be provided", not as "it is already provided elsewhere, and can be reused with no extra work". So, it doesn't solve the problem. We might want to implement something like

```py
any_html: Reuse[BrowserHtml] | HttpResponse
response_url: Reuse[ResponseUrl] | None
```

It seems this would allow to express what we want here cleanly: base class can have a definition like this:

```py
request_or_response_url: Reuse[ResponseUrl] | RequestUrl
```

So, that'd be my proposal:

1. For the current PR just use RequestUrl everywhere.
2. Think about adding Reuse support, to andi / web-poet / scrapy-poet.

What do you think?

---

There is one issue with RequestUrl, which is related to testing (cc @wRAR). Let's say we have a page object which sets url using self.response.url. We generate test fixtures for it using scrapy-poet. Then, we refactor it to use a base class defined here, and want to run tests to see if it works as before. But it seems it won't run, because dependencies are different - there is no RequestUrl dependency in the fixture. I'm not sure, but it might also apply for ResponseUrl vs HttpResponse. Users may need to do some manual steps in this case, or maybe we can provide some helper script, not sure. 

We may also change the savefixture command to always store RequestUrl dependency, even if it's not required by the page object.

---

What do you think @Gallaecio @wRAR @BurnzZ ?

Reply:
I'm not quite sure on using `RequestUrl` everywhere here since it breaks the definition of **url** for Zyte Data Schemas:

> The URL of the final response, after any redirects.

Though I fully agree with the semantics of:

```
request_or_response_url: Reuse[ResponseUrl] | RequestUrl
```

since it's inline with the description of **url** in our schema:

> In case there is no product data on the page or the page was not reached, the returned "empty" item would still contain `url` field and `metadata` field with `dateDownloaded`.

With this, I think we can remove the `_url: ResponseUrl` dependency from `BasePage` and instead have `_BaseMixin` use something like:

```python
import inspect

class _BaseMixin:

    @field
    def url(self) -> str:
        urls = {}

        for name, dep in inspect.signature(self.__init__).parameters.items():
            if dep.annotation is ResponseUrl:
                urls["response_url"] = getattr(self, name)
            elif dep.annotation is HttpResponse:
                urls["response_url"] = getattr(self, name).url
            elif dep.annotation is RequestUrl:
                urls["request_url"] = getattr(self, name)

        url = urls.get("response_url") or urls.get("request_url")
        return str(url)
```

But I'm not sure if introspection is the way to go.

Reply:
> For the current PR just use RequestUrl everywhere.

I’m OK with this given we will be working on addressing this with the `Reuse` approach.

But I do wonder if we should try to get the initial implementation to be closer to idea, with something in line with what @BurnzZ suggests above.

Reply:
What do you think about th following?

1. Have a base class which depends on RequestUrl, and implements url field using self.request_url
2. Page class (which is a subclass of WebPage) re-implements url field using self.response.url

So, there is no ResponseUrl dependency anywhere, and we don't need to do any inspection. There are 2 downsides:

1. We always use RequestUrl dependency, even if it's not needed. I might actually not be a bad thing, as it may testing tool to work better in case of refactoring, and it should be free to get RequestUrl.
2. If a class uses BrowserHtml, it'd need to re-implement url method again to use ResponseUrl. I think that's also fine. We don't have a provider for BrowserHtml yet, and we'd need to do some work to figure out how ResponseUrl is going to work in this case.

Reply:
https://github.com/zytedata/zyte-common-items/pull/19/commits/965b433d61f45065477d731af88299c873118efc